build: display build details link

crazy-max opened the pull request:
https://github.com/docker/buildx/assets/1951866/41a8935a-438f-48f4-be2c-957a56e80f8f

Display build details links if Docker Desktop build backend is enabled.

I'm using an OSC 8 escape sequence with the target links to be displayed in the terminal. Most of the terminal emulators auto-detect when a URL appears onscreen and allow to conveniently open them (e.g. via Ctrl+click or Cmd+click, or the right click menu). A hyperlink is opened upon encountering an OSC 8 escape sequence with the link. The list of supporting terminal emulators and terminal-based apps can be seen here: https://github.com/Alhadis/OSC8-Adoption/.

Atm it only displays links when build is completed and not for active builds. For active builds we need to put links right before build starts so user can just follow build status in desktop instead of his terminal. Would also need a callback to retrieve the ref when attributes are converted to solve opts.

crazy-max commented on go.mod:
```diff
@@ -26,6 +26,7 @@ require (
 	github.com/moby/sys/mountinfo v0.6.2
 	github.com/moby/sys/signal v0.7.0
 	github.com/morikuni/aec v1.0.0
+	github.com/muesli/termenv v0.15.1
```
I could get rid of this lib and just handle the escape sequence in an util package.

crazy-max replied:
Done

crazy-max commented on build/build.go:
```diff
@@ -769,6 +777,9 @@ func BuildWithResultHandler(ctx context.Context, nodes []builder.Node, opt map[s
 	}
 
+	// print build details links
+	logrus.Printf(desktop.BuildDetailsOutput(buildDetailsOpts))
```
As discussed, we should probably output to progress instead of logrus.

tonistiigi replied:
Progress output should be in beginning (and needs more testing with bake). After build this could be a "build summary" section. I think logrus is fine for now and we can turn it into proper summary in next iteration.

tonistiigi replied:
I think it is better if this logging is in `cmd` packages.

jedevc commented on util/desktop/desktop.go:
```diff
@@ -0,0 +1,54 @@
+		if err != nil {
+			return
+		}
+		_, err = os.Stat(filepath.Join(home, ".docker", "desktop-build", ".lock"))
```
Think we should be able to get the right filename using https://github.com/docker/buildx/blob/341fb65f6f568696ed5e2585eb747aedaed7ee1a/util/confutil/config.go#L23 the docker cli?

Or is that inaccurate for some reason?

crazy-max replied:
The desktop build daemon sets the data folder to `$HOME/.docker/desktop-build` and doesn't take into account `DOCKER_CONFIG`.

jedevc commented on util/desktop/desktop.go:
```diff
@@ -0,0 +1,54 @@
+			out.WriteString(fmt.Sprintf("  %s: ", target))
+		}
+		// create an escape sequence using the OSC 8 format: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda
+		out.WriteString(fmt.Sprintf("\033]8;;%s\033\\%s\033]8;;\033\\\n", url, url))
```
We can split this our to a separate function in this file.

tonistiigi commented on util/desktop/desktop.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func hyperlink(url, text string) string {
+	// create an escape sequence using the OSC 8 format: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda
```
Is the hyperlink escape doing anything in here? Iterm works same way without any escape sequences if you just print a link.

If you print escape sequences then you also need to detect that you are actually writing to a terminal, not to a file etc.

crazy-max replied:
Some terminal like Windows Terminal need the escape sequence as it doesn't solve on its own unknown schemes like `docker-desktop://`

tonistiigi commented on build/build.go:
```diff
@@ -736,6 +737,13 @@ func BuildWithResultHandler(ctx context.Context, nodes []builder.Node, opt map[s
+				buildDetailsOpts[k] = desktop.BuildDetailsOpt{
+					Builder: node.Builder,
+					Node:    node.Name,
+					Ref:     so.Ref,
```
We should only print if `Build()` ran and either was successful or we got error from BuildKit. Otherwise BuildKit has not registered the ref yet and it is useless.

Some options:
- Add link details based on status stream. If we get something on status stream then we know that buildkit has registered ref. This is also the time where we would write the link to progress in the future.
- In the `client.Build` error handler wrap the error with a custom error type. Eg. `RemoteBuildError` that does not change the `Error()` method but then caller can check which component returned the error with `errors.Is`.

tonistiigi commented on util/desktop/desktop.go:
```diff
@@ -0,0 +1,54 @@
+
+func BuildDetailsOutput(opts map[string]BuildDetailsOpt) string {
+	var out bytes.Buffer
+	out.WriteString("Open build details on Docker Desktop: ")
```
`s/on/in/`

Maybe `s/Open/Review/` . The shorter we can make this prefix the better.

crazy-max commented on build/build.go:
```diff
@@ -929,6 +933,13 @@ func BuildWithResultHandler(ctx context.Context, nodes []builder.Node, opt map[s
+								}
+							}
 							if resultHandleFunc != nil {
 								resultCtx, err := NewResultContext(ctx, cc, so, res)
```
@jedevc If we want to print in commands pkg I was wondering if we could use the result context handler to set the build details opts?

jedevc replied:
`ResultContext` doesn't leave outside the controller boundary, so that would likely only work for the `runBasicBuild` case I think :cry: Although, we do record the build options, so `controller.Inspect` could work for this?

Though that's only the controller options, not the full solve request.

crazy-max replied:
Ah right thanks for the feedback, I will take a closer look.

crazy-max replied:
Builder name, node name when driver pairs are computed during build are not inputs so doesn't look good to have it in build options. Same for the `refID` computed for each solve opt.

crazy-max replied:
Opened https://github.com/docker/buildx/pull/1861 to carry changes from #1680. Will revisit changes here once switching bake to controller for building.

tonistiigi commented:
@crazy-max Looks like multiple CI issues.

tonistiigi commented:
These comments have not been resolved:


> I think it is better if this logging is in cmd packages.

> If you print escape sequences then you also need to detect that you are actually writing to a terminal, not to a file etc. (see full previous comment for alternatives)

> We should only print if Build() ran and either was successful or we got error from BuildKit. Otherwise BuildKit has not registered the ref yet and it is useless.

The link is not shown at all on error anymore iiuc.

I'm not sure if we should show the link with `--print`. If not then these builds should be marked as internal so they don't get registered by history API.

crazy-max commented:
> I think it is better if this logging is in cmd packages.

Moved to commands package

> If you print escape sequences then you also need to detect that you are actually writing to a terminal, not to a file etc. (see full previous comment for alternatives)

Added detection using `containerd/console`.

> We should only print if Build() ran and either was successful or we got error from BuildKit. Otherwise BuildKit has not registered the ref yet and it is useless.

It now prints on error as well. I was looking at using `ResultContext` and it works fine with `build` command but we don't have any implementation yet for `bake` so in the meantime I return an extra arg that contains the build details.

crazy-max commented on build/build.go:
```diff
@@ -663,18 +664,18 @@ func toSolveOpt(ctx context.Context, node builder.Node, multiDriver bool, opt Op
 }
 
-func Build(ctx context.Context, nodes []builder.Node, opt map[string]Options, docker *dockerutil.Client, configDir string, w progress.Writer) (resp map[string]*client.SolveResponse, err error) {
+func Build(ctx context.Context, nodes []builder.Node, opt map[string]Options, docker *dockerutil.Client, configDir string, w progress.Writer) (resp map[string]*client.SolveResponse, bd map[string]desktop.BuildDetails, err error) {
```
Now using progress to set build detail and print `onclose` in commands package for build and bake.

tonistiigi commented on commands/build.go:
```diff
@@ -614,10 +620,15 @@ func decodeExporterResponse(exporterResponse map[string]string) map[string]inter
 		return nil
 	}
+	if ebr, ok := err.(*build.ErrorWithBuildRef); ok {
+		return errors.Errorf("%v\n\n%s", ebr.Err, buildDetailsOutput(map[string]string{
```
wrap function should not flatten the typed error into a string. This should happen when printing the error, currently handled in main, eg. same as the `errdefs.Sources()` handling.
